Network/Converter/PacketConverter.cs

```diff
@@ -27,321 +28,680 @@
-                return (Packet)Activator.CreateInstance(packetType);
-            else return (Packet)FormatterServices.GetUninitializedObject(packetType);
+            MemoryStream memoryStream =
+                new MemoryStream(serialisedPacket, 0, serialisedPacket.Length);
+
+            BinaryReader binaryReader =
+                new BinaryReader(memoryStream);
+
+            // temporary object whose properties will be set during deserialisation
+            P packet = PacketConverterHelper.InstantiateGenericPacket<P>();
+
+            DeserialiseObjectFromReader(packet, binaryReader);
+
+            return packet;
         }
 
+        #endregion Implementation of IPacketConverter
+
+        #region Serialisation
+
         /// <summary>
-        /// Writes the length of the given array onto the given stream.
-        /// Reads the array out of the property and calls the <see cref="GetBytesFromCustomObject"/> to write the object data.
+        /// Serialises all the properties on the given <see cref="object"/> that
+        /// need to be serialised to the given <see cref="BinaryWriter"/>s
+        /// underlying <see cref="MemoryStream"/>.
         /// </summary>
-        /// <param name="obj">The object.</param>
-        /// <param name="propertyInfo">The property information.</param>
-        /// <param name="binaryWriter">The binary writer.</param>
-        /// <returns>System.Byte[].</returns>
-        private void GetBytesFromArray(object obj, PropertyInfo propertyInfo, BinaryWriter binaryWriter)
+        /// <param name="obj">
+        /// The <see cref="object"/> whose properties to serialise using the given
+        /// <see cref="BinaryWriter"/>.
+        /// </param>
+        /// <param name="binaryWriter">
+        /// The <see cref="BinaryWriter"/> to whose underlying <see cref="MemoryStream"/>
+        /// to serialise the properties of the given <see cref="object"/>.
+        /// </param>
+        /// <remarks>
+        /// This method can only serialise properties that lack the custom
+        /// <see cref="PacketIgnorePropertyAttribute"/>.
+        /// </remarks>
+        private void SerialiseObjectToWriter(object obj, BinaryWriter binaryWriter)
         {
-            Type arrayType = propertyInfo.PropertyType.GetElementType();
-            Array propertyData = (Array)propertyInfo.GetValue(obj);
-            binaryWriter.Write(propertyData?.Length ?? 0);
+            PropertyInfo[] propertiesToSerialise = GetTypeProperties(obj.GetType());
 
-            if (arrayType.IsClass && !IsPrimitive(arrayType)) propertyData.GetEnumerator().ToList<object>().ForEach(p => GetBytesFromCustomObject(p, binaryWriter));
-            else propertyData.GetEnumerator().ToList<object>().ForEach(p => { dynamic targetType = p; binaryWriter.Write(targetType); });
+            for (int i = 0; i < propertiesToSerialise.Length; ++i)
```

Comment:
i++ instead of ++i
There is also a difference between those. Not sure if this is even running correctly

Reply:
++i is actually * slightly * faster. THis is due to an inherent optimisation, with no intermediate value being assigned. 'i++' generates an intermediate value and then assigns it back to 'i'. '++i' increments 'i' directly (i need to check), so is very slightly more efficient. Am willing change back though.

Reply:
I was wrong
It is not faster. 
will change